### PR TITLE
Universal runner scripts Take 2

### PR DESCRIPTION
**Pull request by MattGal**
Contains previous PR changes that were reverted, plus fixes for execution on Mac and local runs.  
Tested with full run on Mac.  The previous Centos issues appear to have been related to not passing in TestNugetRuntimeId property and not my changes.

@weshaggard @ellismg @ericstj

### CODE REVIEW

File: src/Microsoft.DotNet.Build.Tasks/GenerateTestExecutionScripts.cs [hunk 1]
```diff
@@ -57,8 +57,11 @@ private void WriteShExecutionScript(string shExecutionTemplate, string outputPat
                 bool? useAbsolutePath = dependency.GetMetadata("UseAbsolutePath")?.Equals("true", StringComparison.OrdinalIgnoreCase);
                 if (useAbsolutePath == true)
                 {
-                    copyCommands.Append($"cp -l -n \"{dependency.GetMetadata("SourcePath")}\" \"$EXECUTION_DIR/{Path.GetFileName(relativePath)}\" || exit $?\n");
+                    string filePath = dependency.GetMetadata("SourcePath");
+                    string fileName = Path.GetFileName(filePath);
+                    copyCommands.Append($"[ -f $EXECUTION_DIR/{fileName} ] && echo \"{fileName} already exists, not copying.\" || ln {filePath} $EXECUTION_DIR  || exit $?\n");
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/GenerateTestExecutionScripts.cs]: Does exit code of non-zero always mean "file already exists"?

**Comment by ericstj** [comment]
LGTM, would be nice to see CoreFx CI results on a private build but that's not trivial to set up.  Just in case you get adventurous you could build some custom version like 1.0.25-prerelease-00330-01-mattgal, push that to a myget feed, update CoreFx to use that as well as the feed, and submit a PR with "DO NOT MERGE" tag.

**Comment by ellismg** [comment]
> LGTM, would be nice to see CoreFx CI results on a private build but that's not trivial to set up.

Made even worse due to the fact that we will need updates to our CI jobs to be able to consume this, which I am working on.

**Comment by ericstj** [comment]
So once this is merged folks are blocked by the CI job changes?  Any way to put all of this behind a flag so if other folks need buildtools updates they won't be blocked by that?

**Comment by ellismg** [comment]
> So once this is merged folks are blocked by the CI job changes? Any way to put all of this behind a flag so if other folks need buildtools updates they won't be blocked by that?

It would be easier to just wait, I think.  I am working on the updates now, probably done tomorrow.

**Comment by MattGal** [comment]
I agree with @ellismg ,  let's just not accept the PR until he thinks the CI system is ready.

**Comment by weshaggard** [comment]
`copyCommands.Append($"[ -f $EXECUTION_DIR/{fileName} ] && echo \"{fileName} already exists, not copying.\" || ln {filePath} $EXECUTION_DIR  || exit $?\n");`

I'm still interested in understanding this change better.

**Comment by ellismg** [comment]
> I'm still interested in understanding this change better.

For the CoreFX side, we have a requirement that we are able to override the runtime and native libraries that are used (we do this in the cross building case where we need to copy the runtime and native shims over instead of using the runtime and shims restored from packages, since they would be for a different RID).

**Comment by weshaggard** [comment]
OK. So I suspect my understanding of bash is what is holding me back. So is `[ -f $EXECUTION_DIR/{fileName} ] ` that the way to check if a file exists? If so then I do understand the change better but do we think we should create a function called copyandcheck that does this like we do in the cmd file, instead of repeating it for each file?

**Comment by ellismg** [comment]
> So is `[ -f $EXECUTION_DIR/{fileName} ]` that the way to check if a file exists?

Correct.

**Comment by MattGal** [comment]
@weshaggard  pseudocode here of "[ -f $EXECUTION_DIR/{fileName} ] && echo \"{fileName} already exists, not copying.\" || ln {filePath} $EXECUTION_DIR || exit $?\n" is :

if (the file exists in the destination path) 
{
   echo that it's not copying (same as cp -l -n)
}
else
{
   try { copy the file to the destination path}
   catch (any non-0 return code) { exit the script with this return code} 
}

**Comment by weshaggard** [comment]
Do you guys think it is better to generate this for each copy or to depend on a helper function like we are doing in cmd? I kind of think a helper would be better but there might be some reason I'm not understanding why we don't. 

Otherwise LGTM once we get the CI update to be able to use.

**Comment by MattGal** [comment]
it's not a problem, I was on the fence myself and just went this way because it's just 1 line per call.   Making it a function will improve readability, so will fix that and update the PR.  @ellismg  is still working on issues around this with the CI system so I have time.

**Comment by MattGal** [comment]
Just added a property for not copying .ni.dlls if they are included.  Specifying /p:TestWithoutNativeImages=true will prevent mscorlib.ni.dll (and any others, though I did not see them) from showing up in test execution.

**Comment by ellismg** [comment]
> Just added a property for not copying .ni.dlls if they are included.

Thanks!

**Comment by weshaggard** [comment]
LGTM

**Comment by MattGal** [comment]
@dotnet-bot test this please

**Pull request merged**